File: ams/behaviors/lua/ball.lua
--[[
ball behavior - bouncing ball with velocity

The ball moves according to its velocity and bounces off walls.
Paddle/brick collisions are handled by the game layer.

Properties used:
  ball_launched: bool - whether ball is active (set by game)
  ball_attached_to: string - entity ID to follow when not launched

Config options (from YAML):
  speed: float - initial speed magnitude (default: 300)
  max_speed: float - maximum speed (default: 600)
  speed_increase: float - speed added per brick hit (default: 5)

Usage in YAML:
  behaviors:
    - ball:
        speed: 300
        max_speed: 600
]]

local ball = {}

function ball.on_spawn(entity_id)
    ams.set_prop(entity_id, "ball_launched", false)
    ams.set_prop(entity_id, "ball_speed", ams.get_config(entity_id, "ball", "speed", 300))
end

function ball.on_update(entity_id, dt)
    local launched = ams.get_prop(entity_id, "ball_launched")

    if not launched then
        -- Follow attached entity (paddle)
        local attached_id = ams.get_prop(entity_id, "ball_attached_to")
        if attached_id and ams.is_alive(attached_id) then
            local paddle_x = ams.get_x(attached_id)
            local paddle_width = ams.get_width(attached_id)
            local paddle_y = ams.get_y(attached_id)
            local ball_radius = ams.get_width(entity_id) / 2

            -- Center ball on paddle, above it
            ams.set_x(entity_id, paddle_x + paddle_width / 2 - ball_radius)
            ams.set_y(entity_id, paddle_y - ball_radius * 2 - 5)
        end
        return
    end

    -- Move ball
    local x = ams.get_x(entity_id)
    local y = ams.get_y(entity_id)
    local vx = ams.get_vx(entity_id)
    local vy = ams.get_vy(entity_id)

    x = x + vx * dt
    y = y + vy * dt

    ams.set_x(entity_id, x)
    ams.set_y(entity_id, y)

    -- Bounce off walls
    local width = ams.get_width(entity_id)
    local height = ams.get_height(entity_id)
    local screen_w = ams.get_screen_width()
    local screen_h = ams.get_screen_height()

    -- Left wall
    if x < 0 then
        ams.set_x(entity_id, 0)
        ams.set_vx(entity_id, -vx)
        ams.play_sound("wall_bounce")
    end

    -- Right wall
    if x + width > screen_w then
        ams.set_x(entity_id, screen_w - width)
        ams.set_vx(entity_id, -vx)
        ams.play_sound("wall_bounce")
    end

    -- Top wall
    if y < 0 then
        ams.set_y(entity_id, 0)
        ams.set_vy(entity_id, -vy)
        ams.play_sound("wall_bounce")
    end

    -- Bottom (ball lost) - handled by game layer, just mark property
    if y + height > screen_h then
        ams.set_prop(entity_id, "ball_fell", true)
    end
end

-- Called by game to launch the ball
function ball.launch(entity_id, angle_degrees)
    local speed = ams.get_prop(entity_id, "ball_speed") or 300
    local angle_degrees = angle_degrees or -75  -- Default: mostly up, slightly right

    local angle_rad = angle_degrees * 3.14159 / 180
    local vx = speed * ams.cos(angle_rad)
    local vy = speed * ams.sin(angle_rad)

    ams.set_vx(entity_id, vx)
    ams.set_vy(entity_id, vy)
    ams.set_prop(entity_id, "ball_launched", true)
    ams.set_prop(entity_id, "ball_fell", false)
end

-- Called by game after paddle hit to adjust angle
function ball.bounce_off_paddle(entity_id, paddle_center_x, paddle_width)
    local ball_x = ams.get_x(entity_id)
    local ball_width = ams.get_width(entity_id)
    local ball_center_x = ball_x + ball_width / 2

    -- Calculate offset from paddle center (-1 to 1)
    local offset = (ball_center_x - paddle_center_x) / (paddle_width / 2)
    offset = ams.clamp(offset, -1, 1)

    -- Calculate new angle based on offset
    -- Center hit = straight up (-90), edge hit = 60 degrees from vertical
    local max_angle = 60
    local angle = -90 + offset * max_angle
    local angle_rad = angle * 3.14159 / 180

    -- Get current speed
    local vx = ams.get_vx(entity_id)
    local vy = ams.get_vy(entity_id)
    local speed = ams.sqrt(vx * vx + vy * vy)

    -- Set new velocity
    ams.set_vx(entity_id, speed * ams.cos(angle_rad))
    ams.set_vy(entity_id, speed * ams.sin(angle_rad))

    ams.play_sound("paddle_hit")
end

-- Called by game to increase speed after brick hit
function ball.increase_speed(entity_id)
    local speed_increase = ams.get_config(entity_id, "ball", "speed_increase", 5)
    local max_speed = ams.get_config(entity_id, "ball", "max_speed", 600)

    local vx = ams.get_vx(entity_id)
    local vy = ams.get_vy(entity_id)
    local current_speed = ams.sqrt(vx * vx + vy * vy)

    if current_speed >= max_speed then
        return
    end

    local new_speed = math.min(current_speed + speed_increase, max_speed)
    local ratio = new_speed / current_speed

    ams.set_vx(entity_id, vx * ratio)
    ams.set_vy(entity_id, vy * ratio)

    ams.set_prop(entity_id, "ball_speed", new_speed)
end

return ball

File: games/BrickBreaker/game/entities/ball.py
"""Ball entity with velocity-based physics.

The ball bounces off walls, paddle, and bricks. Ball angle depends
on where it hits the paddle.
"""

from dataclasses import dataclass
import math
from typing import Tuple, Optional


@dataclass
class BallConfig:
    """Ball configuration from level or defaults."""

    radius: float = 10.0
    speed: float = 300.0          # Initial speed in pixels/second
    max_speed: float = 600.0
    speed_increase: float = 5.0   # Speed increase per brick hit


class Ball:
    """Ball with velocity-based movement and bouncing physics."""

    def __init__(
        self,
        config: BallConfig,
        x: float,
        y: float,
        vx: float = 0.0,
        vy: float = 0.0,
        active: bool = True,
    ):
        """Initialize ball.

        Args:
            config: Ball configuration
            x: Center X position
            y: Center Y position
            vx: X velocity (pixels/second)
            vy: Y velocity (pixels/second)
            active: Whether ball is active (in play)
        """
        self._config = config
        self._x = x
        self._y = y
        self._vx = vx
        self._vy = vy
        self._active = active

    @property
    def x(self) -> float:
        """Get ball center X."""
        return self._x

    @property
    def y(self) -> float:
        """Get ball center Y."""
        return self._y

    @property
    def vx(self) -> float:
        """Get X velocity."""
        return self._vx

    @property
    def vy(self) -> float:
        """Get Y velocity."""
        return self._vy

    @property
    def radius(self) -> float:
        """Get ball radius."""
        return self._config.radius

    @property
    def speed(self) -> float:
        """Get current ball speed."""
        return math.sqrt(self._vx**2 + self._vy**2)

    @property
    def is_active(self) -> bool:
        """Check if ball is active (in play)."""
        return self._active

    def launch(self, angle_degrees: float = -60.0) -> 'Ball':
        """Launch ball at specified angle.

        Args:
            angle_degrees: Launch angle (0 = right, -90 = up)

        Returns:
            New Ball with velocity set
        """
        angle_rad = math.radians(angle_degrees)
        vx = self._config.speed * math.cos(angle_rad)
        vy = self._config.speed * math.sin(angle_rad)
        return Ball(self._config, self._x, self._y, vx, vy, self._active)

    def update(self, dt: float) -> 'Ball':
        """Update ball position based on velocity.

        Args:
            dt: Delta time in seconds

        Returns:
            New Ball with updated position
        """
        if not self._active:
            return self

        new_x = self._x + self._vx * dt
        new_y = self._y + self._vy * dt

        return Ball(self._config, new_x, new_y, self._vx, self._vy, self._active)

    def set_position(self, x: float, y: float) -> 'Ball':
        """Set ball position.

        Args:
            x: New X position
            y: New Y position

        Returns:
            New Ball at new position
        """
        return Ball(self._config, x, y, self._vx, self._vy, self._active)

    def bounce_horizontal(self) -> 'Ball':
        """Bounce off vertical surface (reverse X velocity).

        Returns:
            New Ball with reversed X velocity
        """
        return Ball(self._config, self._x, self._y, -self._vx, self._vy, self._active)

    def bounce_vertical(self) -> 'Ball':
        """Bounce off horizontal surface (reverse Y velocity).

        Returns:
            New Ball with reversed Y velocity
        """
        return Ball(self._config, self._x, self._y, self._vx, -self._vy, self._active)

    def bounce_off_paddle(self, paddle_x: float, paddle_width: float) -> 'Ball':
        """Bounce off paddle with angle based on hit position.

        Ball bounces at different angles depending on where it hits
        the paddle. Hitting the center bounces straight up, edges
        bounce at steeper angles.

        Args:
            paddle_x: Paddle center X position
            paddle_width: Paddle width

        Returns:
            New Ball with new velocity based on paddle hit position
        """
        # Calculate offset from paddle center (-1 to 1)
        offset = (self._x - paddle_x) / (paddle_width / 2)
        offset = max(-1.0, min(1.0, offset))  # Clamp

        # Calculate bounce angle
        # offset=0 (center) -> -90 degrees (straight up)
        # offset=-1 (left edge) -> -135 degrees (up-left)
        # offset=+1 (right edge) -> -45 degrees (up-right)
        max_angle = 60.0  # Max deviation from vertical (degrees)
        angle_degrees = -90.0 + (offset * max_angle)
        angle_rad = math.radians(angle_degrees)

        # Maintain current speed
        current_speed = self.speed
        vx = current_speed * math.cos(angle_rad)
        vy = current_speed * math.sin(angle_rad)

        # Ensure ball is moving up
        if vy > 0:
            vy = -vy

        return Ball(self._config, self._x, self._y, vx, vy, self._active)

    def increase_speed(self) -> 'Ball':
        """Increase ball speed after hitting a brick.

        Returns:
            New Ball with increased speed
        """
        current_speed = self.speed
        if current_speed >= self._config.max_speed:
            return self

        # Calculate new speed
        new_speed = min(
            current_speed + self._config.speed_increase,
            self._config.max_speed
        )

        # Scale velocity to new speed
        if current_speed > 0:
            scale = new_speed / current_speed
            new_vx = self._vx * scale
            new_vy = self._vy * scale
        else:
            new_vx = self._vx
            new_vy = self._vy

        return Ball(self._config, self._x, self._y, new_vx, new_vy, self._active)

    def deactivate(self) -> 'Ball':
        """Mark ball as inactive (lost).

        Returns:
            New inactive Ball
        """
        return Ball(self._config, self._x, self._y, self._vx, self._vy, False)

    def get_bounds(self) -> Tuple[float, float, float, float]:
        """Get ball bounding box (left, top, right, bottom).

        Returns:
            Tuple of (left, top, right, bottom)
        """
        return (
            self._x - self._config.radius,
            self._y - self._config.radius,
            self._x + self._config.radius,
            self._y + self._config.radius,
        )
